[Impeller] Implement framebuffer-fetch via subpasses in Vulkan without extensions.

chinmaygarde opened the pull request:
* Subpasses are not exposed in the HAL and the need for subpasses in Vulkan can
  be determined based on the presence and use of input-attachments in the
  shaders. This information is already reflected by the compiler. Because of
  this, all references to subpasses have been removed from APIs above the HAL.
* `RenderPassBuilderVK` is a lightweight object used to generate render passes
  to use either with the pipelines (compat, base, or per-subpass) or during
  rendering along with the framebuffer. Using the builder also sets up the
  right subpass dependencies. As long as the builder contains compatible
  attachments and subpass counts, different subpasses stamped by the builder
  (via the `Build` method) are guaranteed to be compatible per the rules in the
  spec.
* Pass attachments are now in the `eGeneral` layout. There was no observable
  difference in performance when manually inserting the right transitions.
  Except, a lot of transitions needed to be inserted. If we need it, we can add
  it back in short order. I wouldn't be averse to adding it if reviewers
  insist.
* Additional pipeline state objects need not be created as the sub-pass
  self-dependencies are sufficient to setup the render-pass.
* Speaking of the `rasterization_order_attachment_access` extension, its use has
  been removed in this patch. I am prototyping adding it back to measure the
  overhead introduced by manual subpass management. If the overhead is
  measurable, we can use the extension on devices that have it as an added
  optimization.
* The complexity of command encoding remains linear (to the number of commands)
  per pass.
* This patch only works on a single color attachment being used as an input
  attachment. While this is sufficient for current use cases, the Metal
  implementation is significantly more capable since the multiple attachments
  and attachment types (depth) are already supported. Rounding out support for
  this is in progress.

chinmaygarde commented:
I was able to reproduce and fix the clips breaking as reported in https://github.com/flutter/engine/pull/49787#issuecomment-1894070685. It was due to a stencil being cleared that should no longer be an issue in this patch. I haven't been able to reproduce the issue reported on a Samsung Android phone because I don't have access to one.

jonahwilliams commented on impeller/renderer/backend/vulkan/formats_vk.h:
```diff
@@ -664,6 +570,51 @@ constexpr vk::ImageAspectFlags ToImageAspectFlags(PixelFormat format) {
   FML_UNREACHABLE();
 }
 
+struct SubpassCursorVK {
```
Is this used anywhere?

chinmaygarde replied:
`PipelineVK::Create` uses a subpass cursor since it could technically still create those subpasses. But we don't call it. I'm backing it out.

chinmaygarde replied:
Done.

jonahwilliams commented:
Awesome, checking this out - quick pass with a few questions.

jonahwilliams commented:
This is working for me locally with a pixel and an older galaxy

jonahwilliams commented on impeller/renderer/backend/vulkan/render_pass_builder_vk.h:
```diff
@@ -0,0 +1,56 @@
+  std::optional<vk::AttachmentDescription> depth_stencil_;
+};
+
+void InsertBarrierForInputAttachmentRead(const vk::CommandBuffer& buffer,
+                                         const vk::Image& image);
```
This method doesn't really seem related to the render_pass builder. Since this is only used in render_pass_vk.cc please move it there

chinmaygarde replied:
Thats not accurate. The access masks and dependency flags must be a subset of the masks and flags specified during subpass self-dependency setup. Those masked are just a few lines above in the implementation. I wanted to keep these two together because the barrier you insert will be closely tied to the self dependency setup, in `render_pass_builder.cc`. I suppose I can make those use a common constant to make that clearer. And also document this either way.

chinmaygarde replied:
I documented and used constants as well adding more context.

jonahwilliams commented on impeller/renderer/backend/vulkan/render_pass_vk.cc:
```diff
@@ -20,23 +20,14 @@
-// manually changed.
-//
-// See: impeller/entity/shaders/blending/framebuffer_blend.frag
-static constexpr size_t kMagicSubpassInputBinding = 64;
```
Please restore

chinmaygarde replied:
Done.

jonahwilliams commented on impeller/renderer/backend/vulkan/render_pass_vk.cc:
```diff
@@ -79,176 +70,65 @@ static std::vector<vk::ClearValue> GetVKClearValues(
-  resolve_refs.resize(render_target_.GetMaxColorAttacmentBindIndex() + 1u,
-                      kUnusedAttachmentReference);
+    const std::shared_ptr<CommandBufferVK>& command_buffer) const {
+  BarrierVK barrier;
```
Unconditionally adding barrier SGTM, the old code was making less and less sense.

jonahwilliams commented on impeller/renderer/renderer_unittests.cc:
```diff
@@ -1300,6 +1307,208 @@ TEST_P(RendererTest,
+  return pipeline;
+}
+
+TEST_P(RendererTest, CanSepiaToneWithSubpasses) {
```
AFAIK (@gaaclarke correct me) but this test file doesn't have goldens right? Do we have sufficient golden coverage in aiks?

chinmaygarde replied:
Good call. Do we have goldens coverage for any advanced blend already?

jonahwilliams replied:
Since we depend on host testing, I'm not sure if anything in our matrix supports fb fetch. Metal desktop only supports it on arm.

Maybe since this path is unconditional swiftshader will still use it and the advanced blends will be covered?

chinmaygarde replied:
The unconditionality of these tests mean that these will only run on devices that support framebuffer-fetch. I forgot to add a skip otherwise. But yeah, on CI, these tests will only work on Vulkan and ARM Macs.

jonahwilliams replied:
but these tests as written are playground tests that aren't golden, so they don't run on CI AFAIK

gaaclarke replied:
It may run, but the test is more like ["does it compile", "does it run without crashing"].  I'm not sure about "passes validation" and we know "results are correct" isn't checked.  Making sure that `impeller_golden_tests` is executing it makes sure we have the later 2 checks.

jonahwilliams reviewed (commented):
This is really exciting!

jonahwilliams commented:
Clips still appear broken when I run the wonderous app on either a Pixel 7 or a Samsung

![flutter_02](https://github.com/flutter/engine/assets/8975114/3497c9bb-eb20-4bc6-975e-82112a2c4bb4)

chinmaygarde commented on impeller/renderer/backend/vulkan/allocator_vk.cc:
```diff
@@ -188,9 +186,7 @@ static constexpr vk::ImageUsageFlags ToVKImageUsageFlags(
-    if (supports_framebuffer_fetch) {
-      vk_usage |= vk::ImageUsageFlagBits::eInputAttachment;
-    }
+    vk_usage |= vk::ImageUsageFlagBits::eInputAttachment;
```
You can use subpassLoad on any image. If a depth-stencil image is set in the descriptor, that will work too as long as there is sufficient synchronization. So, no. Though I just removed this code since supports_framebuffer_fetch will always be true in this code.

chinmaygarde commented on impeller/renderer/backend/vulkan/render_pass_vk.cc:
```diff
@@ -548,6 +428,13 @@ fml::Status RenderPassVK::Draw() {
   );
 
+  if (pipeline_uses_input_attachments_) {
+    const auto& color_attachment0 = TextureVK::Cast(
+        *GetRenderTarget().GetColorAttachments().begin()->second.texture);
+    InsertBarrierForInputAttachmentRead(command_buffer_vk_,
+                                        color_attachment0.GetImage());
```
Yup. Fixed.

chinmaygarde commented:
okie dokie, the test failures have been addressed, the clips work now too and are covered by the golden tests and I think I've addressed all PR comments. PTAL.

jonahwilliams reviewed (approved):
Overall LGTM

I would prefer it if the additional test coverage either added some goldens or something that we could catch on CI.

jonahwilliams commented:
Work towards https://github.com/flutter/flutter/issues/120223

jonahwilliams commented:
Also word of warning, if you force push skia gold won't rerun. I would probably update branch at least once and let it run to verify that there are no regressions.

chinmaygarde commented:
I updated the branch and the goldens seem to be happy. Assuming everything is good. Landing this and working on the new goldens next.
